### PR TITLE
feat: Allow copying an entire session

### LINKED ISSUE #87: Allow copying an entire session
- We already have a [`<CopyButton />`](https://github.com/fmaclen/hollama/blob/main/src/routes/sessions/%5Bid%5D/CopyButton.svelte) component.
- The content of the session can be extracted [from here](https://github.com/fmaclen/hollama/blob/main/src/routes/sessions/%5Bid%5D/%2Bpage.svelte#L39).
- We should add the new instance of `<CopyButton />` [inside this nav](https://github.com/fmaclen/hollama/blob/main/src/routes/sessions/%5Bid%5D/%2Bpage.svelte#L157-L163).

### PR DESCRIPTION
**Pull request by web2ls**
Closes #87

### CODE REVIEW

**Comment by fmaclen** [comment]
@web2ls I'm not sure if you are still working on this PR of it's ready for review but regardless this feature will need a test.

We run integration tests using [Playwright](https://playwright.dev/). I think we can just add new assertions to the [existing "Copy" tests](https://github.com/fmaclen/hollama/blob/main/tests/session.test.ts#L215-L245).

After you are done feel free to ping me or "request" my review.

**Comment by web2ls** [comment]
@fmaclen please check it

File: src/routes/sessions/[id]/+page.svelte [hunk 1]
```diff
@@ -177,6 +190,7 @@
 
 		<svelte:fragment slot="nav">
 			{#if !isNewSession}
+				<CopyButton content={sessionContent} title="Copy session" />
```
Comment (fmaclen) [inline src/routes/sessions/[id]/+page.svelte:193]: I think we can simplify the implementation significantly by doing this:

```suggestion
				<CopyButton content={JSON.stringify(session.messages, null, 2)} />
```

Then we can get rid of the `copySessionContent()` function and the `sessionContent` variable.

The result will retain the structure of the session which is useful if you are pasting this session into new session as context or another LLM entirely:

```json
[
  {
    "role": "user",
    "content": "mic check, 1 2 3"
  },
  {
    "role": "ai",
    "content": "Checking the microphone input, 1...2...3."
  },
  {
    "role": "user",
    "content": "hello hello"
  },
  {
    "role": "ai",
    "content": "Hello! Is there anything I can assist you with today? If you have any questions or need assistance, please feel free to ask."
  }
]
```

File: src/routes/sessions/[id]/CopyButton.svelte [hunk 2]
```diff
@@ -3,14 +3,15 @@
...
+	export let title: string;
 
 	function copyContent() {
 		navigator.clipboard.writeText(content);
 	}
 </script>
 
 <span class="copy-button">
-	<Button title="Copy" variant="icon" size="icon" on:click={copyContent}>
+	<Button title={title} variant="icon" size="icon" on:click={copyContent}>
```
Comment (fmaclen) [inline src/routes/sessions/[id]/CopyButton.svelte:14]: Let's keep the default `tile="Copy"` attribute for now.

File: tests/session.test.ts [hunk 3]
```diff
@@ -237,13 +237,31 @@ test.describe('Session', () => {
...
 	});
 
+	test('can copy the whole session content to clipboard', async ({ page }) => {
+		await page.goto('/');
+		await page.evaluate(() => navigator.clipboard.writeText(""));
+		await chooseModelFromSettings(page, MOCK_API_TAGS_RESPONSE.models[0].name);
+		await page.getByText('Sessions', { exact: true }).click();
+		await mockCompletionResponse(page, MOCK_SESSION_1_RESPONSE_1);
+		await page.getByTestId('new-session').click();
+		await expect(page.getByTitle('Copy session')).toHaveCount(0);
```
Comment (fmaclen) [inline tests/session.test.ts:254]: After removing the custom `title` attribute you can target the "Copy session" button with:

```typescript
await expect(page.locator(".header").getByTitle('Copy')).toHaveCount(0);
```

**Review by fmaclen — CHANGES_REQUESTED** [review]
Thanks for putting together this PR.
It's a great start 👍 

Left you some notes on a few ways we can simplify the implementation.

File: tests/session.test.ts (hunk 3 above)
Comment (fmaclen) [reply tests/session.test.ts:254]: Also, minor formatting issue, can you add an empty line after this assertion?
I'd like to group interactions + assertions in "blocks" separated by empty lines.

**Comment by fmaclen** [comment]
I just merged #90 and there are 2 minor conflicts in your branch.
Let me know if you need help settling them.

Also, this PR is almost done and I would like to merge it soon if you can, if you can't that's ok, just give me a heads up.

**Comment by web2ls** [comment]
@fmaclen 
sorry for delay
I've made changes based on your comments.

**Review by fmaclen — APPROVED** [review]
@web2ls this looks great, thank you!

**Pull request merged**